macros/src/parser/lifetimes.rs

```diff
@@ -0,0 +1,98 @@
+use quote::quote;
+use std::ops::Deref;
+use syn::{parse, spanned::Spanned, GenericArgument, Lifetime, PathArguments, Type};
+
+#[derive(Debug, Clone)]
+pub struct Lifetimes(Vec<Lifetime>);
+
+impl Lifetimes {
+    pub fn new() -> Lifetimes {
+        Lifetimes(Vec::new())
+    }
+
+    pub fn from_type(data_type: &Type) -> Result<Lifetimes, parse::Error> {
+        let mut lifetimes = Lifetimes::new();
+        get_lifetimes(data_type, &mut lifetimes)?;
+        Ok(lifetimes)
+    }
+
+    pub fn push(&mut self, lifetime: &Lifetime) {
+        if !self.0.contains(lifetime) {
+            self.0.push(lifetime.to_owned());
+        }
+    }
+
+    pub fn append(&mut self, lifetimes: &Lifetimes) {
+        for lifetime in lifetimes.0.iter() {
+            self.push(lifetime);
+        }
+    }
+
+    pub fn remove(&mut self, lifetimes: &Lifetimes) {
+        self.0.retain(|lt| !lifetimes.0.contains(lt));
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn to_code(&self) -> proc_macro2::TokenStream {
+        if self.is_empty() {
+            quote! {}
+        } else {
+            let lifetimes = &self.0;
+            quote! { #(#lifetimes),* ,}
+        }
+    }
+}
+
+impl Default for Lifetimes {
+    fn default() -> Self {
+        Lifetimes::new()
+    }
+}
+
+impl Deref for Lifetimes {
+    type Target = [Lifetime];
+
+    fn deref(&self) -> &[Lifetime] {
+        &self.0[..]
+    }
+}
+
+// helper function for extracting lifetimes from a Type
+fn get_lifetimes(data_type: &Type, lifetimes: &mut Lifetimes) -> Result<(), parse::Error> {
```

Comment:
Can we just put this in the function above? I don't see a reason to have this live outside of that function.

Reply:
`get_lifetimes` is recursive for tuples, that is why i left it split.

Reply:
Recursion or not shouldn't really impact whether this is a member func of `Lifetimes`. It seems like this is only called with a `&mut Lifetimes` object, so why not make it a member func that takes `&mut self`?

Reply:
Done! I also renamed it to `insert_from_type`